File: webapp/src/egp/pf.ts
/**
 * Fully adapted particle filter for the log-likelihood of a quantized sample.
 *
 * Hand-synced TypeScript port of egp/pf.py, implementing Section 4.3 of the
 * paper. The hidden state is the window of the last L−1
 * innovations; given that window, X_t is N(μ, h₀²), so the observation weight
 * is the Gaussian cell probability Φ(r) − Φ(l) and the optimal proposal for
 * the new innovation is N(0, 1) truncated to [l, r). Both are available in
 * closed form, which makes the filter fully adapted: resampling uses the
 * exact predictive weights and no importance weights survive the step.
 *
 * All internal quantities are in nats; callers convert to bits.
 */
import { logNdtr, ndtriExp } from './normal'
import type { Rng } from './rng'

const LOG2 = Math.log(2)

/** log(1 − eˣ) for x ≤ 0, accurate at both ends. */
export function log1mexp(x: number): number {
  return x > -LOG2 ? Math.log(-Math.expm1(x)) : Math.log1p(-Math.exp(x))
}

function logaddexp(a: number, b: number): number {
  if (a === -Infinity) return b
  if (b === -Infinity) return a
  const m = Math.max(a, b)
  return m + Math.log(Math.exp(a - m) + Math.exp(b - m))
}

/**
 * log(Φ(hi) − Φ(lo)) for hi ≥ lo, stable deep into either tail.
 *
 * Reflecting an interval whose midpoint is positive keeps the pair of
 * endpoints on the side where logNdtr retains relative precision; without it,
 * two large positive endpoints both give logNdtr ≈ 0 and their difference
 * underflows.
 */
export function logCellProb(lo: number, hi: number): number {
  const mirror = lo + hi > 0
  const a = mirror ? -hi : lo
  const b = mirror ? -lo : hi
  const la = logNdtr(a)
  const lb = logNdtr(b)
  return lb + log1mexp(la - lb)
}

export const DEFAULT_COLLAPSE_MARGIN = 25

/**
 * Fraction of lost-lock steps above which a replicate is unusable. A filter
 * that is tracking has none at all, so this is a wide margin around zero
 * rather than a tuned cutoff. Such a replicate's likelihood is an arbitrary
 * number, not a noisy one, so it must be discarded rather than averaged in.
 */
export const LOST_LOCK_TOLERANCE = 1e-3

/**
 * Per-step log-likelihood below which the filter has certainly lost the state.
 *
 * A particle sitting d prediction standard deviations from the observed cell
 * contributes about log(Δ/h₀) − d²/2, so a filter that is tracking at all
 * cannot fall far below log(min(1, Δ/h₀)). Anything `margin` nats under that
 * is a lost-lock step, not an unlucky symbol. The scale term matters: under
 * fine quantization *every* step is legitimately worth about log(Δ/h₀) nats,
 * so a fixed threshold would misfire. (ESS does not detect this failure —
 * the surviving particles agree with each other whether or not they agree
 * with the data.)
 */
export function collapseThreshold(delta: number, h0: number, margin = DEFAULT_COLLAPSE_MARGIN): number {
  return Math.log(Math.min(1, delta / h0)) - margin
}

export interface FilterResult {
  loglikNats: number
  nSteps: number
  /** −loglik / (n ln 2): the entropy-rate estimate in bits per sample. */
  entropyRateBits: number
  /** Effective sample size, averaged over steps, as a fraction of N. */
  meanEss: number
  minEss: number
  nLostLock: number
  /** −1 if the filter never lost lock. */
  firstLostStep: number
  lostLockFraction: number
  collapseThreshold: number
}

/**
 * Estimate log P(y₁…yₙ) for the quantized FIR process.
 *
 * `taps` must be the minimum-phase factorization with taps[0] > 0;
 * `y[t] = round(x[t] / delta)`. For a white process (L = 1) the likelihood is
 * exact and `nParticles` is ignored.
 */
export function particleFilterLoglik(
  y: Int32Array,
  taps: Float64Array,
  delta: number,
  nParticles: number,
  rng: Rng,
  collapseMargin = DEFAULT_COLLAPSE_MARGIN,
  progress?: (done: number, total: number) => void,
): FilterResult {
  const n = y.length
  const nTaps = taps.length
  const h0 = taps[0]
  if (!(h0 > 0)) throw new Error('taps[0] must be positive (use the minimum-phase factorization)')
  if (!(delta > 0)) throw new Error('delta must be positive')
  const limit = collapseThreshold(delta, h0, collapseMargin)
  const width = delta / h0

  const finish = (loglik: number, meanEss: number, minEss: number, nLost: number, firstLost: number): FilterResult => ({
    loglikNats: loglik,
    nSteps: n,
    entropyRateBits: n > 0 ? -loglik / (n * LOG2) : NaN,
    meanEss,
    minEss,
    nLostLock: nLost,
    firstLostStep: firstLost,
    lostLockFraction: n > 0 ? nLost / n : 0,
    collapseThreshold: limit,
  })

  if (n === 0) return finish(0, 1, 1, 0, -1)

  if (nTaps === 1) {
    // White: symbols are i.i.d., the likelihood is exact.
    let loglik = 0
    let nLost = 0
    let firstLost = -1
    for (let t = 0; t < n; t++) {
      const lo = ((y[t] - 0.5) * delta) / h0
      const step = logCellProb(lo, lo + width)
      loglik += step
      if (step < limit) {
        nLost++
        if (firstLost < 0) firstLost = t
      }
    }
    return finish(loglik, 1, 1, nLost, firstLost)
  }

  const N = nParticles | 0
  if (N < 2) throw new Error('nParticles must be >= 2')

  const ringLen = nTaps - 1
  // Particle-major ring buffer of the last L−1 innovations. The tap vector is
  // rolled each step (tapAt below) so that μ is a plain contiguous dot product
  // per particle — the O(nNL) inner loop of the whole estimator.
  let ring = new Float64Array(N * ringLen)
  for (let i = 0; i < ring.length; i++) ring[i] = rng.normal()
  let scratch = new Float64Array(N * ringLen)
  const tapAt = new Float64Array(ringLen)

  const lo = new Float64Array(N)
  const logAlpha = new Float64Array(N)
  const logLower = new Float64Array(N)
  const mirror = new Uint8Array(N)
  const cumulative = new Float64Array(N)
  const ancestors = new Int32Array(N)

  let pos = 0
  let loglik = 0
  let essSum = 0
  let essMin = N
  let nLost = 0
  let firstLost = -1
  const logN = Math.log(N)
  const reportEvery = Math.max(1, Math.floor(n / 100))

  for (let t = 0; t < n; t++) {
    // Rolled tap vector so ring[·, (pos − k) mod ringLen] pairs with taps[1+k].
    for (let k = 0; k < ringLen; k++) {
      tapAt[(((pos - k) % ringLen) + ringLen) % ringLen] = taps[1 + k]
    }
    const lower = (y[t] - 0.5) * delta

    // Weights: the Gaussian cell probability per particle, in log space. The
    // weight and the optimal proposal share all their transcendentals.
    let peak = -Infinity
    for (let i = 0; i < N; i++) {
      let mu = 0
      const base = i * ringLen
      for (let j = 0; j < ringLen; j++) mu += ring[base + j] * tapAt[j]
      const l = (lower - mu) / h0
      const h = l + width
      const flip = l + h > 0
      const a = flip ? -h : l
      const b = flip ? -l : h
      const la = logNdtr(a)
      const lb = logNdtr(b)
      lo[i] = l
      logAlpha[i] = lb + log1mexp(la - lb)
      logLower[i] = la
      mirror[i] = flip ? 1 : 0
      if (logAlpha[i] > peak) peak = logAlpha[i]
    }
    if (!Number.isFinite(peak)) {
      throw new Error(
        `particle filter collapsed at step ${t}: every particle assigns zero ` +
          'probability to the observed symbol. Increase the particle count.',
      )
    }

    let wSum = 0
    let wSumSq = 0
    for (let i = 0; i < N; i++) {
      const w = Math.exp(logAlpha[i] - peak)
      wSum += w
      cumulative[i] = wSum
      wSumSq += w * w
    }
    const step = peak + Math.log(wSum) - logN
    loglik += step
    if (step < limit) {
      nLost++
      if (firstLost < 0) firstLost = t
    }
    const ess = (wSum * wSum) / wSumSq
    essSum += ess
    if (ess < essMin) essMin = ess

    // Systematic resampling with the exact predictive weights; the stratified
    // u's are increasing, so one merge pointer replaces per-particle search.
    const u0 = rng.uniform()
    let j = 0
    for (let i = 0; i < N; i++) {
      const u = ((u0 + i) / N) * wSum
      while (j < N - 1 && cumulative[j] < u) j++
      ancestors[i] = j
    }
    for (let i = 0; i < N; i++) {
      const src = ancestors[i] * ringLen
      const dst = i * ringLen
      for (let k = 0; k < ringLen; k++) scratch[dst + k] = ring[src + k]
    }
    const prev = ring
    ring = scratch
    scratch = prev

    // Propagate: the optimal proposal is the truncated innovation, drawn by
    // inverse CDF in log space from the ancestor's precomputed cell stats.
    pos = (pos + 1) % ringLen
    for (let i = 0; i < N; i++) {
      const a = ancestors[i]
      const u = rng.uniform()
      const logP = logaddexp(logLower[a], Math.log(u) + logAlpha[a])
      let x = ndtriExp(logP)
      if (mirror[a]) x = -x
      const l = lo[a]
      const h = l + width
      if (!Number.isFinite(x)) x = 0.5 * (l + h) // numerically empty cell
      ring[i * ringLen + pos] = x < l ? l : x > h ? h : x
    }

    if (progress && (t % reportEvery === 0 || t === n - 1)) progress(t + 1, n)
  }

  return finish(loglik, essSum / (n * N), essMin / N, nLost, firstLost)
}

File: egp/src/egp/pf.py
"""Fully adapted particle filter for the log-likelihood of a quantized sample.

Implements Section 4.3 of the paper.  The hidden state is the
window of the last L-1 innovations; given that window, X_t is N(mu, h0^2), so
the observation weight is the Gaussian cell probability Phi(r) - Phi(l) and the
optimal proposal for the new innovation is N(0, 1) truncated to [l, r).  Both
are available in closed form, which makes the filter fully adapted: resampling
uses the exact predictive weights and no importance weights survive the step.

All internal quantities are in nats; callers convert to bits.
"""

from __future__ import annotations

from dataclasses import dataclass

import numpy as np
from scipy.special import log_ndtr

try:  # SciPy >= 1.8
    from scipy.special import ndtri_exp
except ImportError:  # pragma: no cover - fallback for very old SciPy
    from scipy.special import ndtri

    def ndtri_exp(logp):
        return ndtri(np.exp(logp))


_LOG2 = np.log(2.0)


def _log1mexp(x) -> np.ndarray:
    """log(1 - exp(x)) for x <= 0, accurate at both ends."""
    x = np.asarray(x, dtype=float)
    flat = np.atleast_1d(x)
    out = np.empty_like(flat)
    near = flat > -_LOG2
    with np.errstate(divide="ignore", invalid="ignore"):
        out[near] = np.log(-np.expm1(flat[near]))
        far = ~near
        out[far] = np.log1p(-np.exp(flat[far]))
    return out.reshape(x.shape)


def _cell_stats(lo: np.ndarray, hi: np.ndarray):
    """Shared work behind both the weight and the proposal.

    Returns ``(log_prob, log_lower_cdf, mirror)`` where ``log_prob`` is
    ``log(Phi(hi) - Phi(lo))`` and the other two are what the inverse-CDF draw
    needs.  Reflecting an interval whose midpoint is positive keeps the pair of
    endpoints on the side where ``log_ndtr`` retains relative precision; without
    it, two large positive endpoints both give ``log_ndtr ~ 0`` and their
    difference underflows.
    """
    mirror = (lo + hi) > 0
    a = np.where(mirror, -hi, lo)
    b = np.where(mirror, -lo, hi)
    la = log_ndtr(a)
    lb = log_ndtr(b)
    return lb + _log1mexp(la - lb), la, mirror


def log_cell_prob(lo, hi) -> np.ndarray:
    """log(Phi(hi) - Phi(lo)) for hi >= lo, stable deep into either tail."""
    return _cell_stats(np.asarray(lo, dtype=float), np.asarray(hi, dtype=float))[0]


def _draw(log_mass, la, mirror, lo, hi, rng: np.random.Generator) -> np.ndarray:
    """Inverse-CDF draw from N(0, 1) on [lo, hi) given precomputed cell stats."""
    u = rng.random(la.shape)
    with np.errstate(divide="ignore"):
        log_p = np.logaddexp(la, np.log(u) + log_mass)
    x = ndtri_exp(log_p)
    x = np.where(mirror, -x, x)
    bad = ~np.isfinite(x)
    if bad.any():  # numerically empty cell: fall back to its midpoint
        x = np.where(bad, 0.5 * (lo + hi), x)
    return np.clip(x, lo, hi)


def sample_truncnorm(lo: np.ndarray, hi: np.ndarray, rng: np.random.Generator) -> np.ndarray:
    """Draw N(0, 1) truncated to [lo, hi) by inverse CDF in log space."""
    lo = np.asarray(lo, dtype=float)
    hi = np.asarray(hi, dtype=float)
    log_mass, la, mirror = _cell_stats(lo, hi)
    return _draw(log_mass, la, mirror, lo, hi, rng)


DEFAULT_COLLAPSE_MARGIN = 25.0

#: Fraction of lost-lock steps above which a replicate is unusable.  A filter
#: that is tracking has none at all, so this is a wide margin around zero
#: rather than a tuned cutoff.
LOST_LOCK_TOLERANCE = 1e-3


@dataclass
class FilterResult:
    loglik_nats: float  # over all n steps
    loglik_counted: float  # over the steps after burn-in
    n_steps: int
    burn_in: int
    mean_ess: float  # averaged effective sample size, as a fraction of N
    min_ess: float
    n_lost_lock: int
    first_lost_step: int  # -1 if the filter never lost lock
    collapse_threshold: float

    @property
    def n_counted(self) -> int:
        return self.n_steps - self.burn_in

    @property
    def entropy_rate_bits(self) -> float:
        if self.n_counted <= 0:
            return float("nan")
        return -self.loglik_counted / (self.n_counted * _LOG2)

    @property
    def lost_lock_fraction(self) -> float:
        return self.n_lost_lock / self.n_steps if self.n_steps else 0.0

    @property
    def lost_lock(self) -> bool:
        """True when the particle cloud lost the state on this run.

        The log-likelihood of such a run is not a noisy estimate of the true
        one but an arbitrary number, so callers must discard it rather than
        average it in.
        """
        return self.lost_lock_fraction > LOST_LOCK_TOLERANCE


def collapse_threshold(delta: float, h0: float, margin: float = DEFAULT_COLLAPSE_MARGIN) -> float:
    """Per-step log-likelihood below which the filter has certainly lost the state.

    A particle sitting ``d`` prediction standard deviations away from the
    observed cell contributes about ``log(delta / h0) - d^2 / 2``, so a filter
    that is tracking at all cannot fall far below ``log(min(1, delta / h0))``.
    Anything ``margin`` nats under that is a lost-lock step, not an unlucky
    symbol.  The scale term matters: under fine quantization *every* step is
    legitimately worth about ``log(delta / h0)`` nats.
    """
    return float(np.log(min(1.0, delta / h0)) - margin)


def particle_filter_loglik(
    y,
    taps,
    delta: float,
    n_particles: int,
    rng: np.random.Generator,
    burn_in: int = 0,
    collapse_margin: float = DEFAULT_COLLAPSE_MARGIN,
    progress=None,
) -> FilterResult:
    """Estimate log P(y_1..y_n) for the quantized FIR process.

    Parameters
    ----------
    y : integer array
        Observed symbols, ``y_t = round(x_t / delta)``.
    taps : array
        Minimum-phase FIR taps; ``taps[0]`` must be positive.
    delta : float
        Quantization step.
    n_particles : int
        Number of particles N.  Ignored when the process is white (L = 1), for
        which the filter is exact.
    burn_in : int
        Number of leading steps excluded from ``loglik_counted``.  The filter
        still runs them, so the excluded part is what lets the particle cloud
        lock onto the state; dropping it estimates
        ``-log P(y_{B+1..n} | y_1..y_B) / (n - B)``, which is still an upper
        bound on the entropy rate but a tighter one.
    collapse_margin : float
        Nats below :func:`collapse_threshold` at which a step counts as
        lost-lock.  See :func:`collapse_threshold`.
    progress : callable, optional
        Called as ``progress(t, n)`` every ~1% of the sweep.
    """
    y = np.asarray(y, dtype=np.int64).ravel()
    taps = np.asarray(taps, dtype=float).ravel()
    delta = float(delta)
    n = y.size
    n_taps = taps.size
    h0 = float(taps[0])
    if not h0 > 0:
        raise ValueError("taps[0] must be positive (use the minimum-phase factorization)")
    if delta <= 0:
        raise ValueError("delta must be positive")
    burn_in = int(np.clip(burn_in, 0, max(0, n - 1)))
    limit = collapse_threshold(delta, h0, collapse_margin)
    if n == 0:
        return FilterResult(0.0, 0.0, 0, 0, 1.0, 1.0, 0, -1, limit)

    lower = (y - 0.5) * delta
    width = delta / h0

    if n_taps == 1:  # white: symbols are i.i.d., the likelihood is exact
        lo = lower / h0
        steps = log_cell_prob(lo, lo + width)
        return FilterResult(
            loglik_nats=float(np.sum(steps)),
            loglik_counted=float(np.sum(steps[burn_in:])),
            n_steps=n,
            burn_in=burn_in,
            mean_ess=1.0,
            min_ess=1.0,
            n_lost_lock=int(np.sum(steps < limit)),
            first_lost_step=int(np.argmax(steps < limit)) if np.any(steps < limit) else -1,
            collapse_threshold=limit,
        )

    n_particles = int(n_particles)
    if n_particles < 2:
        raise ValueError("n_particles must be >= 2")

    ring_len = n_taps - 1
    ring = rng.standard_normal((n_particles, ring_len))
    scratch = np.empty_like(ring)
    tap_buf = np.empty(ring_len)
    offsets = np.arange(ring_len)
    particle_index = np.arange(n_particles)
    tail = taps[1:]
    pos = 0

    loglik = 0.0
    loglik_counted = 0.0
    ess_sum = 0.0
    ess_min = float(n_particles)
    n_lost = 0
    first_lost = -1
    log_n = np.log(n_particles)
    report_every = max(1, n // 100)

    for t in range(n):
        # Rolled tap vector so that ring[:, (pos - k) % ring_len] holds W_{t-1-k}.
        tap_buf[(pos - offsets) % ring_len] = tail
        mu = ring @ tap_buf

        lo = (lower[t] - mu) / h0
        hi = lo + width
        # The weight and the optimal proposal share all their transcendentals.
        log_alpha, log_lower, mirror = _cell_stats(lo, hi)

        peak = log_alpha.max()
        if not np.isfinite(peak):
            raise RuntimeError(
                f"particle filter collapsed at step {t}: every particle assigns "
                "zero probability to the observed symbol. Increase --particles."
            )
        weights = np.exp(log_alpha - peak)
        w_sum = float(weights.sum())
        step = peak + np.log(w_sum) - log_n
        loglik += step
        if t >= burn_in:
            loglik_counted += step
        if step < limit:
            n_lost += 1
            if first_lost < 0:
                first_lost = t

        ess = w_sum * w_sum / float(np.dot(weights, weights))
        ess_sum += ess
        ess_min = min(ess_min, ess)

        # Systematic resampling with the exact predictive weights.
        cumulative = np.cumsum(weights)
        cumulative /= cumulative[-1]
        u = (rng.random() + particle_index) / n_particles
        ancestors = np.searchsorted(cumulative, u)
        np.clip(ancestors, 0, n_particles - 1, out=ancestors)
        np.take(ring, ancestors, axis=0, out=scratch)
        ring, scratch = scratch, ring

        # Propagate: the optimal proposal is the truncated innovation.
        w_new = _draw(
            log_alpha[ancestors],
            log_lower[ancestors],
            mirror[ancestors],
            lo[ancestors],
            hi[ancestors],
            rng,
        )
        pos = (pos + 1) % ring_len
        ring[:, pos] = w_new

        if progress is not None and (t % report_every == 0 or t == n - 1):
            progress(t + 1, n)

    return FilterResult(
        loglik_nats=float(loglik),
        loglik_counted=float(loglik_counted),
        n_steps=n,
        burn_in=burn_in,
        mean_ess=ess_sum / (n * n_particles),
        min_ess=ess_min / n_particles,
        n_lost_lock=n_lost,
        first_lost_step=first_lost,
        collapse_threshold=limit,
    )
